### PR TITLE
Huge PR to kick off 0.9

### PR DESCRIPTION
**Pull request by jasoncalabrese**
This will pull in all the work @MilosKozak has been doing on the `dev-0.9` branch including the new profile model and bolus calculator.

Hoping to get this merged in soon, and then do some refactoring, and fill in tests after it's on `dev`

### CODE REVIEW

**Comment by MilosKozak** [comment]
wait with tests until the code become more 'final' :)

**Comment by jasoncalabrese** [comment]
yeah, just want to make sure we don't break any pre-existing features before we merge all this in.  Also need to make the profile collection testing situation more clear to everyone.

**Comment by francescaneri** [comment]
I have to finish to translate two things . I hope to give her this evening thanks ;)

File: lib/report_plugins/treatments.js [hunk 1]
```diff
@@ -268,13 +284,14 @@ treatments.report = function report_treatments(datastorage, sorteddaystoshow, op
     .append($('<th>').css('width','50px').attr('align','left').append(translate('Duration')))
     .append($('<th>').css('width','50px').attr('align','left').append(translate('Percent')))
     .append($('<th>').css('width','50px').attr('align','left').append(translate('Basal value')))
+    .append($('<th>').css('width','50px').attr('align','left').append(translate('Profile')))
     .append($('<th>').css('width','150px').attr('align','left').append(translate('Entered By')))
     .append($('<th>').css('width','300px').attr('align','left').append(translate('Notes')))
   );
   
   sorteddaystoshow.forEach(function (day) {
```
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1239128172/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: lib/client/boluscalc.js [hunk 2]
```diff
@@ -0,0 +1,767 @@
...
+      $('#bc_carbsneededtr').css('display','none');
+      $('#bc_insulinneededtr').css('display','');
+    }
+
+    // Show basal rate
+    var basal = client.sbx.data.profile.getTempBasal(record.eventTime);
+    $('#bc_basal').text((basal.treatment ? 'T ' : '') + basal.tempbasal.toFixed(3));
+  };
+  
+  boluscalc.gatherBoluscalcData = function gatherBoluscalcData() {
```
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1239128117/issues/source?bid=2452379&fileBranchId=2935247#l312)

File: lib/client/renderer.js [hunk 3]
```diff
@@ -621,6 +646,73 @@ function init (client, d3) {
...
+    treatProfiles.enter().append('text')
+      .attr('class', 'g-profile')
+      .style('font-size', 15)
+      .style('font-weight', 'bold')
+      .attr('fill', '#0099ff')
+      .attr('text-anchor', 'end')
+      .attr('dy', '.35em')
+      .attr('transform', function (t) {
+        return 'rotate(-90 ' + chart().xScale(t.mills) + ',' + chart().yScaleBasals(topOfText) + ') ' +
+      'translate(' + chart().xScale(t.mills) + ',' + chart().yScaleBasals(topOfText) + ')'
```
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1239128123/issues/source?bid=2452379&fileBranchId=2935247#l700)

File: lib/client/boluscalc.js [hunk 4]
```diff
@@ -0,0 +1,767 @@
...
+  };
+  
+  boluscalc.calculateInsulin = function calculateInsulin (event) {
+    maybePrevent(event);
+    boluscalc.gatherBoluscalcData( );
+    boluscalc.updateGui(boluscalc.record);
+    return  boluscalc.record;
+  };
+  
+  boluscalc.updateGui = function updateGui (record) {
```
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1239128117/issues/source?bid=2452379&fileBranchId=2935247#l184)

File: static/profile/js/profileeditor.js [hunk 5]
```diff
@@ -139,59 +155,216 @@
...
-      dateInput.css({'background-color':'green'});
-      submitButton.css({'background-color':'green'});
-    } else {
-      submitButton.css('display','none');
-      timeInput.css({'background-color':'red'});
-      dateInput.css({'background-color':'red'});
-      submitButton.css({'background-color':'red'});
+  function profileAdd (event) {
+    var record = mongorecords[currentrecord];
+    var newname = 'New profile';
```
Comment (jasoncalabrese) [inline static/profile/js/profileeditor.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1239128186/issues/source?bid=2452379&fileBranchId=2935247#l329)

File: lib/report_plugins/treatments.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240173027/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: lib/client/boluscalc.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240172972/issues/source?bid=2452379&fileBranchId=2935247#l312)

File: lib/client/renderer.js (hunk 3 above)
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240172978/issues/source?bid=2452379&fileBranchId=2935247#l700)

File: lib/client/boluscalc.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240172972/issues/source?bid=2452379&fileBranchId=2935247#l184)

File: static/profile/js/profileeditor.js (hunk 5 above)
Comment (jasoncalabrese) [inline static/profile/js/profileeditor.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240173041/issues/source?bid=2452379&fileBranchId=2935247#l329)

File: lib/report_plugins/treatments.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240289755/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: lib/client/boluscalc.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240289700/issues/source?bid=2452379&fileBranchId=2935247#l312)

File: lib/client/renderer.js (hunk 3 above)
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240289706/issues/source?bid=2452379&fileBranchId=2935247#l700)

File: lib/client/boluscalc.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240289700/issues/source?bid=2452379&fileBranchId=2935247#l184)

File: static/profile/js/profileeditor.js (hunk 5 above)
Comment (jasoncalabrese) [inline static/profile/js/profileeditor.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1240289769/issues/source?bid=2452379&fileBranchId=2935247#l329)

File: lib/report_plugins/treatments.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242270524/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: static/profile/js/profileeditor.js [hunk 6]
```diff
@@ -139,59 +155,216 @@
...
+    }
+
+    maybePreventDefault(event);
+    return false;
+  }
+  
+  function profileClone (event) {
+    GUIToObject();
+    var record = mongorecords[currentrecord];
+    var newname = $('#pe_profile_name').val() + ' (copy)';
```
Comment (jasoncalabrese) [inline static/profile/js/profileeditor.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242270538/issues/source?bid=2452379&fileBranchId=2935247#l359)

File: lib/client/boluscalc.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242270469/issues/source?bid=2452379&fileBranchId=2935247#l312)

File: lib/client/boluscalc.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242270469/issues/source?bid=2452379&fileBranchId=2935247#l184)

File: lib/client/boluscalc.js [hunk 7]
```diff
@@ -0,0 +1,767 @@
...
+    
+    // Carbs needed if too much iob
+    record.carbsneeded = 0;
+    if (record.insulin<0) {
+      record.carbsneeded = Math.round(-total * ic);
+    }
+    
+    console.log('Insulin calculation result: ',record);
+    return record;
+  }
```
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242270469/issues/source?bid=2452379&fileBranchId=2935247#l445)

File: lib/report_plugins/treatments.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242327822/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: lib/client/boluscalc.js [hunk 8]
```diff
@@ -0,0 +1,767 @@
...
+  
+  boluscalc.submit = function submit(event) {
+    var data = gatherData();
+    if (data) {
+      confirmPost(data);
+    }
+    maybePrevent(event);
+    return false;
+  }
```
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242327767/issues/source?bid=2452379&fileBranchId=2935247#l487)

File: lib/client/boluscalc.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242327767/issues/source?bid=2452379&fileBranchId=2935247#l312)

File: lib/client/boluscalc.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242327767/issues/source?bid=2452379&fileBranchId=2935247#l184)

File: lib/report_plugins/treatments.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242348859/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: lib/client/boluscalc.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242348804/issues/source?bid=2452379&fileBranchId=2935247#l312)

File: lib/client/boluscalc.js [hunk 9]
```diff
@@ -0,0 +1,767 @@
...
+            categories[r.category][r.subcategory] = true;
+          }
+        });
+        databaseloaded = true;
+        console.log('Food database loaded');
+        fillForm();
+      }
+    }).done(function() { if (callback) { callback(); } });
+    maybePrevent(event);
+  }
```
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242348804/issues/source?bid=2452379&fileBranchId=2935247#l616)

File: lib/client/boluscalc.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242348804/issues/source?bid=2452379&fileBranchId=2935247#l184)

File: lib/report_plugins/treatments.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242642325/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: lib/client/boluscalc.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242642270/issues/source?bid=2452379&fileBranchId=2935247#l312)

File: lib/client/boluscalc.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1242642270/issues/source?bid=2452379&fileBranchId=2935247#l184)

**Comment by jasoncalabrese** [comment]
### BWP and Bolus Calc don't always agree...   Using different targets?

#### BWP = 0 (not < 0)
screen shot 2015-11-02 at 6 34 21 pm

#### Bolus Calc  = Needs 2g carbs -0.15
screen shot 2015-11-02 at 6 34 40 pm

**Comment by MilosKozak** [comment]
boluscalc still offers carbs equivalent (or insulin correction) even if calculation fit into target range. usualy it's very small correction you don't react to. will think about what's better.
we liked to know there such a correction possible


your example perfect fit to this situation: child is asking for a candy (3g usualy). I see 2g carbs needed, arrow down so we give candy and don't care about insulin
so bwp is correct = no action needed but boluscalc makes our decision easier

**Comment by MilosKozak** [comment]
we need more opinions to this... others?

**Comment by francescaneri** [comment]
Profile editor fuller . The only thing is that if you keep on the same database two versions of the 0.8.2 and 0.9.0 will remain the profile editor of 0.9.0 , while that of 0.8.2 is canceled

**Comment by MilosKozak** [comment]
@francescaneri you have to set different MONGO_PROFILE_COLLECTION if you plan to use 0.8.2 and 0.9 together

**Comment by francescaneri** [comment]
I did not know.  food editor : you should be able to give a starting database to try to standardize / work together to standardize and make more precise the bolus calculator .

**Comment by StuArnold** [comment]
@francescaneri I was thinking the same about a food database - either the ability to load a database or  perhaps a shared central database by country (names/ carb values may vary between countries) that you can subscribe to. Would need somebody to populate it though! In the larger groups I can see that being shared out - everybody including those entering the data would benefit by having to enter less than individually

**Comment by MilosKozak** [comment]
well i don't like to have databese full of foods i don't use. it makes it harder to select. but if there will be such wish we can add some tool

File: lib/report_plugins/treatments.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1247894710/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: lib/client/boluscalc.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1247894655/issues/source?bid=2452379&fileBranchId=2935247#l312)

File: lib/client/boluscalc.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1247894655/issues/source?bid=2452379&fileBranchId=2935247#l184)

**Comment by MilosKozak** [comment]
I analyzed the difference BWP vs boluscalc and found it's very specific situation. Solved by adding this
image

**Comment by jasoncalabrese** [comment]
Think that makes it even more complicated, if you're in the target range why show the carbs to bring you to the top of the target range

**Comment by MilosKozak** [comment]
ok. 
image
i'll never see it anyway because we don't use range but a single value.
adjusted some rouding to be closer to BWP but some difference could sometimes appear because of some rounding in the middle of calculation which is not the same

File: lib/client/boluscalc.js [hunk 10]
```diff
@@ -0,0 +1,776 @@
...
+      $('#bc_insulinneededtr').css('display','');
+      $('#bc_calculationintarget').css('display','none');
+    }
+
+    // Show basal rate
+    var basal = client.sbx.data.profile.getTempBasal(record.eventTime);
+    $('#bc_basal').text((basal.treatment ? 'T: ' : '') + basal.tempbasal.toFixed(3));
+  };
+  
+  boluscalc.gatherBoluscalcData = function gatherBoluscalcData() {
```
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1249279887/issues/source?bid=2452379&fileBranchId=2935247#l321)

File: lib/report_plugins/treatments.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1249279942/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: lib/client/boluscalc.js [hunk 11]
```diff
@@ -0,0 +1,776 @@
...
+  };
+  
+  boluscalc.calculateInsulin = function calculateInsulin (event) {
+    maybePrevent(event);
+    boluscalc.gatherBoluscalcData( );
+    boluscalc.updateGui(boluscalc.record);
+    return  boluscalc.record;
+  };
+  
+  boluscalc.updateGui = function updateGui (record) {
```
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1249279887/issues/source?bid=2452379&fileBranchId=2935247#l184)

File: lib/client/boluscalc.js (hunk 10 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (20)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1270838365/issues/source?bid=2452379&fileBranchId=2935247#l321)

File: lib/report_plugins/treatments.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/report_plugins/treatments.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1270838421/issues/source?bid=2452379&fileBranchId=2935247#l292)

File: lib/client/boluscalc.js (hunk 11 above)
Comment (jasoncalabrese) [inline lib/client/boluscalc.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1270838365/issues/source?bid=2452379&fileBranchId=2935247#l184)

**Pull request merged**